include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,288 @@
+  ~tracking_resource_adaptor()                                = default;
+  tracking_resource_adaptor(tracking_resource_adaptor const&) = delete;
+  tracking_resource_adaptor(tracking_resource_adaptor&&)      = default;
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor const&) = delete;
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor&&) = default;
+
+  /**
+   * @brief Return pointer to the upstream resource.
+   *
+   * @return Upstream* Pointer to the upstream resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @brief Checks whether the upstream resource supports streams.
+   *
+   * @return true The upstream resource supports streams
+   * @return false The upstream resource does not support streams.
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+  /**
+   * @brief Get the outstanding allocations map
+   *
+   * @return std::map<void*, allocation_info> const& of a map of allocations. The key
+   * is the allocated memory pointer and the data is the allocation_info structure, which
+   * contains size and, potentially, stack traces.
+   */
+  std::map<void*, allocation_info> const& get_outstanding_allocations() const
+  {
+    return allocations;
+  }
+
+  /**
+   * @brief Query the number of bytes that have been allocated. Note that
+   * this can not be used to know how large of an allocation is possible due
+   * to both possible fragmentation and also internal page sizes and alignment
+   * that is not tracked by this allocator.
+   *
+   * @return std::size_t number of bytes that have been allocated through this
+   * allocator.
+   */
+  std::size_t get_allocated_bytes() const { return allocated_bytes_; }
+
+  /**
+   * @brief Log any outstanding allocations via RMM_LOG_DEBUG
+   *
+   */
+  void log_outstanding_allocations() const
+  {
+#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
+    read_lock_t lock(mtx);
+    if (not allocations.empty()) {
+      std::ostringstream oss;
+      for (auto const& al : allocations) {
+        oss << al.first << ": " << al.second.allocation_size << " B";
+        if (al.second.strace != nullptr) {
+          oss << " : callstack:" << std::endl << *al.second.strace;
+        }
+        oss << std::endl;
+      }
+      RMM_LOG_DEBUG("Outstanding Allocations: {}", oss.str());
+    }
+#endif  // SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
+  }
+
+  /**
+   * @brief Get the number of outstanding allocations
+   *
+   * @return std::size_t number of allocations still outstanding
+   */
+  std::size_t get_num_outstanding_allocations() const { return allocations.size(); };
```

Comment:
Is this method necessary? It's the same as calling `get_outstanding_allocations().size()`.

```suggestion
  std::size_t get_num_outstanding_allocations() const noexcept { return allocations.size(); };
```

Reply:
This was added before we had `get_outstanding_allocations`. I don't think it is necessary any longer.